compression_verifier.go

```diff
@@ -0,0 +1,238 @@
+type UnsupportedCompressionError struct {
+	table     string
+	column    string
+	algorithm string
+}
+
+func (e UnsupportedCompressionError) Error() string {
+	return "Compression algorithm: " + e.algorithm +
+		" not supported on table: " + e.table +
+		" for column: " + e.column
+}
+
+// CompressionVerifier provides support for verifying the payload of compressed columns that
+// may have different hashes for the same data by first decompressing the compressed
+// data before fingerprinting
+type CompressionVerifier struct {
+	logger *logrus.Entry
+
+	// supportedAlgorithms provide O(1) lookup to check if a configured algorithm is supported
+	supportedAlgorithms map[string]struct{}
+
+	// tableColumnCompressions is represented as table -> column -> compression-type
+	tableColumnCompressions TableColumnCompressionConfig
+}
+
+// NewCompressionVerifier first checks the map for supported compression algorithms before
+// initializing and returning the initialized instance.
+func NewCompressionVerifier(tableColumnCompressions TableColumnCompressionConfig) (*CompressionVerifier, error) {
+	supportedAlgorithms := make(map[string]struct{})
+	supportedAlgorithms[CompressionSnappy] = struct{}{}
+
+	compressionVerifier := &CompressionVerifier{
+		logger:                  logrus.WithField("tag", "compression_verifier"),
+		supportedAlgorithms:     supportedAlgorithms,
+		tableColumnCompressions: tableColumnCompressions,
+	}
+
+	if err := compressionVerifier.verifyConfiguredCompression(tableColumnCompressions); err != nil {
+		return nil, err
+	}
+
+	return compressionVerifier, nil
+}
+
+func (c *CompressionVerifier) verifyConfiguredCompression(tableColumnCompressions TableColumnCompressionConfig) error {
+	for table, columns := range tableColumnCompressions {
+		for column, algorithm := range columns {
+			// Ensure algorithm is uppercased
+			algorithm = strings.ToUpper(algorithm)
+			tableColumnCompressions[table][column] = algorithm
+
+			if _, ok := c.supportedAlgorithms[algorithm]; !ok {
+				return &UnsupportedCompressionError{
+					table:     table,
+					column:    column,
+					algorithm: algorithm,
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// GetCompressedHashes compares the source data with the target data to ensure the integrity of the
+// data being copied.
+//
+// The GetCompressedHashes method checks if the existing table contains compressed data
+// and will apply the decompression algorithm to the applicable columns if necessary.
+// After the columns are decompressed, the hashes of the data are used to verify equality
+func (c *CompressionVerifier) GetCompressedHashes(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64, tableCompression map[string]string) (map[uint64][]byte, error) {
+	c.logger.Info("decompressing table data before verification")
+
+	// Extract the raw rows using SQL to be decompressed
+	rows, err := c.getRows(db, schema, table, pkColumn, columns, pks)
```

Comment:
I think a lot of confusion would be cleared up if the `getRows` function did the hard work of massaging `SELECT pk_col, *` to `map[pk]rowData` where `rowData`'s indexes map one-to-one with the `columns` slice.